pkg/manager/manager_test.go

```diff
@@ -0,0 +1,110 @@
+package manager
+
+import (
+	"fmt"
+	"os"
+	"testing"
+	"time"
+
+	. "github.com/onsi/gomega"
+	"github.com/pkg/errors"
+	"gopkg.in/fsnotify.v1"
+
+	"sigs.k8s.io/cluster-api-provider-vsphere/pkg/context/fake"
+)
+
+func TestManager_FileWatch(t *testing.T) {
+	g := NewWithT(t)
+	contentFmt := `---
+username: '%s'
+password: '%s'
+`
+
+	tests := []struct {
+		name            string
+		username        string
+		updatedUsername string
+		password        string
+		updatedPassword string
+		delFile         bool
+		failWatch       bool
+	}{
+		{
+			name:            "update username & password for CAPV credentials",
+			username:        "abcd",
+			updatedUsername: "efgh",
+			password:        "pass",
+			updatedPassword: "ssap",
+			delFile:         false,
+		},
+		{
+			name:            "send an error on watch error channel",
+			username:        "abcd",
+			updatedUsername: "efgh",
+			password:        "pass",
+			updatedPassword: "ssap",
+			delFile:         true,
+		},
+		{
+			name:            "force fail the watch",
+			username:        "abcd",
+			updatedUsername: "efgh",
+			password:        "pass",
+			updatedPassword: "ssap",
+			delFile:         false,
+			failWatch:       true,
+		},
+	}
+	for _, tc := range tests {
+		tmpFile, err := os.CreateTemp("", "creds")
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		managerOptsTest := &Options{
+			// needs an object ref to be present
+			CredentialsFile: tmpFile.Name(),
+			Username:        tc.username,
+			Password:        tc.password,
+		}
+
+		t.Run(tc.name, func(t *testing.T) {
+			if tc.failWatch {
+				managerOptsTest.CredentialsFile = "fail"
+			}
+			watch, err := InitializeWatch(fake.NewControllerManagerContext(), managerOptsTest)
+			// Match initial credentials
+			g.Expect(managerOptsTest.Username).To(Equal(tc.username))
+			g.Expect(managerOptsTest.Password).To(Equal(tc.password))
+
+			// Match updated credentials
+			if tc.delFile {
+				watch.Errors <- errors.Errorf("force failure")
+			}
+
+			if tc.failWatch {
+				g.Expect(err).NotTo(BeNil())
+			} else {
+				// Update the file and wait for watch to detect the change
+				content := fmt.Sprintf(contentFmt, tc.updatedUsername, tc.updatedPassword)
+				updateFunction(content, tmpFile)
+
+				// Add buffer time to avoid race condition while updating the credentials
+				time.Sleep(500 * time.Microsecond)
```

Comment:
Not a good idea to add a `Sleep()` in test functions. Use `g.Eventually()` to assert the update of the credentials in the context.